pkgs/by-name/xr/xrsh/package.nix

```
@@ -0,0 +1,62 @@
+{
+  stdenv,
+  fetchFromGitea,
+  makeWrapper,
+  writeShellApplication,
+  python3Packages,
+  lib,
+}:
+
+stdenv.mkDerivation rec {
+  pname = "xrsh";
+  version = "0.143";
+
+  src = fetchFromGitea {
+    fetchSubmodules = true;
+    domain = "codeberg.org";
+    owner = "xrsh";
+    repo = "xrsh";
+    url = "https://codeberg.org/xrsh/xrsh.git";
+    rev = "435ac2fc51d5966a86607bb5ede3661a210157d4";
+    hash = "sha256-g+YxeExomlWHF7suzlPjUlZe6eXa9MPA10bCoifQO1I=";
```

The hash appears to be wrong:

```suggestion
    hash = "sha256-R4xkbZYjyi+TVPhv07c45o4x8AqeZGBS/u8Pw/aZWvc=";
```

This is a common mistake, which usually happens when you forget to update the hash, but since the old one is already cached in your system (becoming a [fixed-output derivation](https://wiki.nixos.org/wiki/Glossary)), the build succeeds and no error is raised.

I usually always set all hashes to an empty string `hash = ""` when updating things so I avoid this issue.

---

OK I've updated the hash, still weird that `nix-prefetch-git` literally gave me that hash:

```
$ nix-prefetch-git https://codeberg.org/xrsh/xrsh.git
Initialized empty Git repository in /tmp/git-checkout-tmp-ddxbUO1v/xrsh/.git/
remote: Enumerating objects: 75, done.
remote: Counting objects: 100% (75/75), done.
remote: Compressing objects: 100% (59/59), done.
remote: Total 75 (delta 5), reused 38 (delta 3), pack-reused 0 (from 0)
Unpacking objects: 100% (75/75), 21.45 MiB | 6.44 MiB/s, done.
From https://codeberg.org/xrsh/xrsh
 * branch            HEAD       -> FETCH_HEAD
Switched to a new branch 'fetchgit'
removing `.git'...

git revision is 435ac2fc51d5966a86607bb5ede3661a210157d4
path is /nix/store/ndgj8358z9rrvd2c986vlsasbkd0h9yy-xrsh
git human-readable version is -- none --
Commit date is 2025-02-21 16:03:10 +0100
hash is 0livs0ks5hj6sz0c7x6swplmwmjjwd9wwbmv2y3mb6k89iw33rl3
{
  "url": "https://codeberg.org/xrsh/xrsh.git",
  "rev": "435ac2fc51d5966a86607bb5ede3661a210157d4",
  "date": "2025-02-21T16:03:10+01:00",
  "path": "/nix/store/ndgj8358z9rrvd2c986vlsasbkd0h9yy-xrsh",
  "sha256": "0livs0ks5hj6sz0c7x6swplmwmjjwd9wwbmv2y3mb6k89iw33rl3",
  "hash": "sha256-g+YxeExomlWHF7suzlPjUlZe6eXa9MPA10bCoifQO1I=",
  "fetchLFS": false,
  "fetchSubmodules": false,
  "deepClone": false,
  "leaveDotGit": false
}
```

---

It would probably be better to not rely on fetchers directly for this exact reason. You'd also want to build the package anyways to make sure it works, so setting the hash to an empty string makes sense.

That said, adding an update script would probably be even better:

```
  passthru.updateScript = unstableGitUpdater { };
```

With this, you just can use [nix-update](https://search.nixos.org/packages?channel=unstable&show=nix-update&from=0&size=50&sort=relevance&type=packages&query=nix-update) and it will take care of updating the source and hash for you:

```
nix-update -u xrsh
```

```diff
-    rev = "435ac2fc51d5966a86607bb5ede3661a210157d4";
-    hash = "sha256-g+YxeExomlWHF7suzlPjUlZe6eXa9MPA10bCoifQO1I=";
+    rev = "5d6e9f4615e57f01ba0859000d90f4f162a06653";
+    hash = "sha256-9I6c1GkO7XQZ242ZBCxPiPxkZaYYi/UDLoUvS6nE+M0=";
```

This also makes it possible for the [update bot](https://github.com/r-ryantm) to automatically check if new versions exist, open a PR for them and ping maintainers for review.

---

Won't the updater rip out the `0.143` base part of the version, since the repo doesn't use tags to base it on & it doesn't know how to extract that information from a tag in an HTML file.

---

Interesting, yup you're right. I suppose we can either change the version to `0-unstable...` or just remove the updater, unless there is a better solution.

---

@coderofsalvation 2 Options here for you:

1. Use the `updateScript` so Nixpkgs automation can bump the package for you. This will either require you to accept that the version of the package will be `0-unstable-YYYY-MM-DD`, or require you to properly tag your versions in your repository so the automation will use the correct version numbers. If you do add tags, then you can also use the `gitUpdater` instead so only stable versions get bumped to.
2. Leave the updating completely up to you.

Up to you which one you want to go with (though my opinion is that an `updateScript` would be the way to go).